File: TinyScript/js/src/test/parser/SimpleParser.test.ts
import Lexer from "../../main/js/lexer/Lexer";
import Token from "../../main/js/lexer/Token";
import arrayToGenerator from "../../main/js/commons/ArrayToGenerator";
import SimpleParser from "../../main/js/parser/SimpleParser";
import PeekTokenIterator from "../../main/js/parser/PeekTokenIterator";
import ASTNode from "../../main/js/parser/ast/ASTNode";
import * as assert from "assert";

describe("test simple parser", () => {
    test("test 1+2+3+4", () => {
        const source: string = "1+2+3+4";
        const lexer: Lexer = new Lexer();
        const tokens: Token[] = lexer.analyse(arrayToGenerator([...source]));
        const iterator = new PeekTokenIterator(arrayToGenerator(tokens));
        const astNode: ASTNode = SimpleParser.parse(iterator);

        expect(astNode.lexeme?.value).toEqual("+")
        const v1 = astNode.getChildren(0) as ASTNode;
        const e1 = astNode.getChildren(1) as ASTNode;

        expect(v1.lexeme?.value).toEqual("1");
        expect(e1.lexeme?.value).toEqual("+");

        const v2 = e1.getChildren(0) as ASTNode;
        const e2 = e1.getChildren(1) as ASTNode;

        expect(v2.lexeme?.value).toEqual("2");
        expect(e2.lexeme?.value).toEqual("+");

        const v3 = e2.getChildren(0) as ASTNode;
        const e3 = e2.getChildren(1) as ASTNode;

        expect(v3.lexeme?.value).toEqual("3");
        expect(e3.lexeme?.value).toEqual("4");
    })
})

File: TinyScript/java/src/test/java/paser/ast/SimpleParserTest.java
package paser.ast;

import lexer.Lexer;
import lexer.LexicalException;
import org.junit.jupiter.api.Test;
import paser.util.ParseException;
import paser.util.PeekTokenIterator;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/9/3 0:30
 **/
class SimpleParserTest {

    @Test
    void test() throws LexicalException, ParseException {
        Stream<Character> source = "1+2+3+4".chars().mapToObj(x -> (char) x);
        Lexer lexer = new Lexer();
        PeekTokenIterator iterator = new PeekTokenIterator(lexer.analyse(source).stream());
        ASTNode expr = SimpleParser.parse(iterator);
        assertEquals(2, expr.getChildren().size());
        Scalar v1 = (Scalar) expr.getChildren(0);
        assertEquals("1", v1.getLexeme().getValue());
        assertEquals("+",expr.getLexeme().getValue());

        var e2 = (Expr) expr.getChildren(1);
        var v2 = (Scalar) e2.getChildren(0);
        assertEquals("2", v2.getLexeme().getValue());
        assertEquals("+", e2.getLexeme().getValue());

        var e3 = (Expr) e2.getChildren(1);
        var v3 = (Scalar) e3.getChildren(0);
        assertEquals("3", v3.getLexeme().getValue());
        assertEquals("+", e3.getLexeme().getValue());

        Scalar v4 = (Scalar) e3.getChildren(1);
        assertEquals("4", v4.getLexeme().getValue());

    }
}